cpp/src/explainer/tree_shap.cu

```diff
@@ -178,82 +552,46 @@ std::unique_ptr<TreePathInfo> extract_path_info_impl(
   if (!std::is_same<ThresholdType, LeafType>::value) {
     RAFT_FAIL("ThresholdType and LeafType must be identical");
   }
-  if (model.task_type != tl::TaskType::kBinaryClfRegr &&
-      model.task_type != tl::TaskType::kMultiClfGrovePerClass) {
-    RAFT_FAIL("cuML RF / scikit-learn classifiers are not yet supported");
+  if (!std::is_same<ThresholdType, float>::value && !std::is_same<ThresholdType, double>::value) {
+    RAFT_FAIL("ThresholdType must be either float32 or float64");
   }
+
   std::unique_ptr<TreePathInfo> path_info_ptr = std::make_unique<TreePathInfoImpl<ThresholdType>>();
   auto* path_info = dynamic_cast<TreePathInfoImpl<ThresholdType>*>(path_info_ptr.get());
 
+  /* 1. Scan the model for categorical splits and pre-allocate bit fields. */
+  CategoricalSplitCounter<ThresholdType, LeafType> cat_counter{model.num_feature};
+  visit_path_segments_in_model(model, cat_counter);
```

Comment:
I love your visitor design pattern to do the counting as well as the path generation.